### PR TITLE
fix(ivy): View Queries inheritance fix

### PR DESCRIPTION
**Pull request by AndrewKushnir**
Prior to this change `viewQuery` functions that represent @ViewQuery list were not composable, which caused problems in case one Component/Directive inherits another one and both of them contain View Queries. Due to the fact that we used indices to reference queries, resulting query set was corrupted (child component queries were overridden by super class ones). In order to avoid that we no longer use indices assigned at compile time and instead maintain current view query index while iterating through them. This allows us to compose `viewQuery` functions and make inheritance feature work with View Queries.

Note: Content Queries have similar issue and it will be resolved within the next PR.

This PR was created based on the changes from the previous one #28304 (which is now closed) and after some additional debugging due to the failing CircleCI task.

This PR resolves FW-829.

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by AndrewKushnir** [comment]
@kara FYI I've addressed Misko's feedback that he left in the previous version of this PR (#28304) and renamed `createViewQuery` -> `viewQuery`.

**Review by kara — CHANGES_REQUESTED** [review]
General approach looks great. Just a bunch of nits.

File: packages/core/src/render3/query.ts [hunk 1]
```diff
@@ -394,3 +388,35 @@ export function queryRefresh(queryList: QueryList<any>): boolean {
   }
   return false;
 }
+
+/**
+ * Creates new QueryList, stores the reference in LView and returns QueryList.
```
Comment (JoostK) [inline packages/core/src/render3/query.ts]: Since this is public api doc, there should be no mentioning of `LView` here.

File: packages/core/src/render3/query.ts [hunk 2]
```diff
@@ -394,3 +388,35 @@ export function queryRefresh(queryList: QueryList<any>): boolean {
...
+  }
+  const index = getCurrentViewQueryIndex();
+  const viewQuery: QueryList<T> = query<T>(predicate, descend, read);
+  store(index, viewQuery);
+  setCurrentViewQueryIndex(index + 1);
+  return viewQuery;
+}
+
+/**
+* Loads current View Query and moves the pointer/index to the next View Query in LView.
```
Comment (JoostK) [inline packages/core/src/render3/query.ts]: Also public api, so no `LView` references here.

File: packages/core/src/render3/query.ts (hunk 1 above)
Comment (JoostK) [reply packages/core/src/render3/query.ts]: > Wait, wait, this is an instruction in the generated code I don't think we consider this public API just yet.

@pkozlowski-opensource Correct, not yet, but aren't all Ivy instructions supposed to be considered "public", even if they are currently privately exported?

Comment (kara) [reply packages/core/src/render3/query.ts]: Not yet, no. @AndrewKushnir I think it's fine as is for now, esp since we have a number of instructions that reference `LView` already.

Comment (pkozlowski-opensource) [reply packages/core/src/render3/query.ts]: > but aren't all Ivy instructions supposed to be considered "public", even if they are currently privately exported?

Not till we decide to ship generated code to npm and / or open up instructions for people so they can create they own generators. Both mentioned things requires substantial preparations and we might decide against if we see major issues. We will have to dive into it at some point but we are not there yet.

tl;dr; is that I wouldn't worry about comments on instructions for now (especially that we've got impl details on many other instructions already).

**Comment by AndrewKushnir** [comment]
Thanks everyone for the review!

@kara I've addressed your comments in https://github.com/angular/angular/pull/28309/commits/1690f8d88204ebcfb8101b3ec320a00f0da4a791 fixup commit. Could you please have another look?

Thank you.

**Review by kara — APPROVED** [review]
LGTM

**Comment by AndrewKushnir** [comment]
[Presubmit](http://test/OCL:230584186:BASE:230584306:1548276762374:fb1b88c7)

**Review by alxhub — APPROVED** [review]
